File: week_1/day-2-prefix-sum-hashing/subarray_sum_k_follow_up_questions/streaming_subarray_sum.js
// streaming_subarray_sum.js

class StreamingSubarraySum {
    constructor(k) {
        this.k = k;
        this.prefixSum = 0;
        this.freq = new Map();
        this.freq.set(0, 1);
        this.count = 0;
    }

    add(num) {
        this.prefixSum += num;
        if (this.freq.has(this.prefixSum - this.k)) {
            this.count += this.freq.get(this.prefixSum - this.k);
        }
        this.freq.set(this.prefixSum, (this.freq.get(this.prefixSum) || 0) + 1);
        return this.count;
    }
}

// Example Usage
const stream = new StreamingSubarraySum(3);
const inputs = [1, 2, 1, 1];

for (let num of inputs) {
    console.log(stream.add(num));
}
// Output:
// 0
// 1
// 1
// 2

File: week_1/day-2-prefix-sum-hashing/subarray_sum_k_follow_up_questions/streaming_subarray_sum.py
# streaming_subarray_sum.py

class StreamingSubarraySum:
    def __init__(self, k):
        self.k = k
        self.prefix_sum = 0
        self.freq = {0: 1}
        self.count = 0

    def add(self, num):
        self.prefix_sum += num
        self.count += self.freq.get(self.prefix_sum - self.k, 0)
        self.freq[self.prefix_sum] = self.freq.get(self.prefix_sum, 0) + 1
        return self.count

# Example Usage
stream = StreamingSubarraySum(3)
inputs = [1, 2, 1, 1]

for num in inputs:
    print(stream.add(num))
# Output:
# 0
# 1
# 1
# 2
